mmpy_bot/plugins/base.py

```diff
@@ -105,4 +127,142 @@ def get_help_string(self):
 
     async def help(self, message: Message):
         """Prints the list of functions registered on every active plugin."""
-        self.driver.reply_to(message, self.get_help_string())
+        if self.direct_help:
+            self.driver.reply_to(message, self.get_help_string(), direct=True)
+        else:
+            self.driver.reply_to(message, self.get_help_string())
+
+
+@dataclass
+class PluginHelp:
+    help_type: str
+    location: str
+    function: str
+    pattern: str
+    doc_header: str
+    doc_full: str
+    direct: bool
+    mention: bool
+    annotations: Dict
+
+
+class PluginManager:
+    """PluginManager is responsible for initializing all plugins and display aggregated
+    help from each of them.
+
+    It is supposed to be transparent to EventHandler that interacts directly with each
+    individual Plugin.
+    """
+
+    def __init__(
+        self,
+        plugins: Sequence[Plugin],
+        help_trigger: bool = True,
+        help_trigger_bang: bool = False,
+        direct_help: bool = True,
+    ):
+        self.driver: Optional[Driver] = None
+        self.plugins: Sequence[Plugin] = plugins
+        self.direct_help: bool = direct_help
+        self.call_function: Optional[Callable] = None
+
+        self.message_listeners: Dict[
+            re.Pattern, Sequence[MessageFunction]
+        ] = defaultdict(list)
+
+        # This code is a bit hairy because the function signature of an
+        # instance is (message) not (self, message) causing failures later
+        if help_trigger:
+            self.help = listen_to("^help$", needs_mention=True)(Plugin.help)
+        if help_trigger_bang:
+            if not help_trigger:
+                self.help = listen_to("^!help$")(Plugin.help)
+            else:
+                self.help = listen_to("^!help$")(self.help)
+
+    def __iter__(self):
+        return iter(self.plugins)
+
+    def initialize(self, driver: Driver, settings: Optional[Settings] = None):
+        self.driver = driver
+        self.call_function = caller(driver)
+
+        # Add Plugin manager help to message listeners
+        self.help.plugin = self
+        self.message_listeners[self.help.matcher].append(self.help)
+
+        for plugin in self.plugins:
+            plugin.initialize(self.driver, settings)
+
+    def get_help(self):
```

Comment:
The code in get_help() seems like just cut and paste with only 3 differences, might be better for maintainability if you refactored this and put the common parts into their own function.

Reply:
Refactored.